core/modules/loader/CentralClient.cc

```diff
@@ -161,10 +164,27 @@ KeyInfoData::Ptr CentralClient::keyInsertReq(std::string const& key, int chunk,
                              " subchunk=" << subchunk);
     auto keyInsertOneShot = std::make_shared<CentralClient::KeyInsertReqOneShot>(this, key, chunk, subchunk);
     {
-        std::lock_guard<std::mutex> lck(_waitingKeyInsertMtx);
+        std::unique_lock<std::mutex> lck(_waitingKeyInsertMtx);
+        // Limit the number of concurrent inserts.
+        // If the key is already in the map, there is no point in blocking.
+        int loopCount = 0;
         auto iter = _waitingKeyInsertMap.find(key);
+        while (_waitingKeyInsertMap.size() > _doListMaxInserts
+               && iter == _waitingKeyInsertMap.end()) {
+            size_t sz = _waitingKeyInsertMap.size();
+            lck.unlock();
+            if (loopCount % 100 == 0) {
+                LOGS(_log, LOG_LVL_INFO, "keyInsertReq waiting key=" << key <<
+                        "size=" << sz << " loopCount=" << loopCount);
+            }
+            usleep(_maxRequestSleepTime);
```

Comment:
Loops like this can burn significant CPU cycles doing nothing useful while waiting for something to finish. A short sleep can significantly reduce wasted cycles and is much simpler than a condition variable solution.